File: Data Structures & Algorithms/minimum-window-with-characters/submission-13.js
class Solution {
    /**
     * @param {string} s
     * @param {string} t
     * @return {string}
     */
    minWindow(s, t) { 

        let count = {}  
        let need = 0 
        let minLength = s.length + 1
        let minStart = 0

        for (const c of t) {
            count[c] = ( count[c] || 0 ) + 1
        }  

        need = Object.values(count).length

        let l = 0 
        for (let r = 0; r < s.length; r ++) {
            let cur = s[r] 
            if (count[cur] !== undefined) {
                count[cur] -- 
                if (count[cur] === 0) {
                    need --
                }
            } 

            while(need === 0) { 
                if (r - l + 1 < minLength) {
                    minLength = r - l + 1;
                    minStart = l;
                }
                let deleted = s[l]  

                if (count[deleted] !== undefined) {
                    count[deleted] ++ 
                    if (count[deleted] > 0) {
                        need ++
                    }
                }  

                l ++ 
            }
        } 

        return minLength === s.length + 1 ? '' : s.substring(minStart, minStart + minLength)
    }
}

File: Data Structures & Algorithms/permutation-string/submission-13.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2): 
            return False 
            
        count = Counter(s1) 

        for i in range(len(s1)):  
            count[s2[i]] -= 1 

        if all( value == 0 for value in count.values() ): 
            return True 

        l = 0
        for r in range(len(s1), len(s2)): 
            cur = s2[r] 
            deleted = s2[l] 
            count[cur] -= 1 
            count[deleted] += 1  
            l += 1

            if all( value == 0 for value in count.values() ): 
                return True 

        return False